src/lib/components/Segment.svelte

```diff
@@ -0,0 +1,121 @@
+<script lang="ts">
+  import { writable } from "svelte/store";
+  import type { SegmentContext, SelectedSegment } from "$lib/types/segment";
+  import { SEGMENT_CONTEXT_KEY } from "$lib/types/segment";
+  import { setContext } from "svelte";
+
+  export let selectedSegmentId: symbol | undefined = undefined;
+
+  const store = writable<SelectedSegment>({
+    id: selectedSegmentId,
+    element: undefined,
+  });
+
+  setContext<SegmentContext>(SEGMENT_CONTEXT_KEY, {
+    store,
+  });
+
+  $: $store, (() => (selectedSegmentId = $store.id))();
+
+  let indicator:
+    | {
+        left: number;
+        width: number;
+      }
+    | undefined = undefined;
+
+  let segment: HTMLElement | undefined | null;
+
+  let selectedElement: HTMLElement | undefined | null;
+  $: selectedElement =
+    $store.element ?? segment?.querySelector(".segment-button");
+
+  export const initIndicator = () => {
+    if (selectedElement === undefined || selectedElement === null) {
+      indicator = undefined;
+      return;
+    }
+
+    const { parentElement } = selectedElement;
+
+    if (parentElement === null) {
+      indicator = undefined;
+      return;
+    }
+
+    const { left: parentClientLeft } = parentElement.getBoundingClientRect();
+    const { left: currentClientLeft, width: currentClientWidth } =
+      selectedElement.getBoundingClientRect();
+
+    indicator = {
+      left: currentClientLeft - parentClientLeft,
+      width: currentClientWidth,
+    };
+  };
+
+  $: selectedElement, (() => initIndicator())();
+
+  let segmentsCount = 0;
+  $: segment,
+    (() =>
+      (segmentsCount =
+        segment?.querySelectorAll(".segment-button").length ?? 0))();
+</script>
+
+<div
+  bind:this={segment}
+  class="segment"
+  style={`--segments: ${segmentsCount}; ${
+    indicator !== undefined
+      ? `--position: ${indicator.left}px; --width: ${indicator.width}px`
+      : ""
+  }`}
+>
+  <div class="indicator">
+    <div class="indicator-background" />
+  </div>
+
+  <slot />
+</div>
+
+<style lang="scss">
+  .segment {
+    display: grid;
+    grid-auto-columns: minmax(0, 1fr);
+    grid-auto-flow: column;
+    align-items: center;
+
+    background: var(--overlay-background);
+    color: var(--overlay-background-contrast);
+
+    padding: var(--padding) var(--padding-2x);
+    gap: var(--padding-2x);
+
+    border-radius: var(--border-radius);
+    overflow: hidden;
+
+    width: var(--segment-min-width, var(--segment-width, fit-content));
+    --segment-min-width: calc(var(--segment-button-width) * var(--segments));
+
+    position: relative;
+  }
+
+  .indicator {
```

Comment:
If I understand correctly, this is the indicator of the current selected element?

Why didn't you add the style in the child instead? You have the element there, wouldn't it be easier?

Reply:
Because the animation is easier to implement with a single indicator `div` that moves along the all segment length (see GIF in https://github.com/dfinity/nns-dapp/pull/1842).

If we add the indicator in each segment button, then each shall moves on click which makes the animation logic a bit more tricky. Not against, if you want to implement it in another PR for sure that's cool.

Reply:
I understand. For the transition that makes sense.